File: All Lab Tasks/Lab 1/RideSharingApp__Strategy_Pattern_/src/Trip.java
public class Trip {
    private int id;
    private String pickupLocation;
    private String dropOffLocation;
    private RideType rideType;
    private double fare;
    private double distance;
    private Driver driver;
    private String status;

    public Trip(int id, String pickupLocation, String dropOffLocation, RideType rideType, double distance) {
        this.id = id;
        this.pickupLocation = pickupLocation;
        this.dropOffLocation = dropOffLocation;
        this.rideType = rideType;
        this.distance = distance;
        this.fare = calculateFare();
        this.status = "Requested";
    }

    public double calculateFare() {
        switch (rideType) {
            case CARPOOL:
                return distance * 20;

            case LUXURY:
                return distance * 35;

            case BIKE:
                return distance * 15;

            default:
                return distance * 10;
        }
    }

    public void assignDriver(Driver driver) {
        this.driver = driver;
        this.status = "Driver Assigned";
    }

    public void startTrip() {
        this.status = "In Progress";
        System.out.println("Trip has started.");
    }

    public void completeTrip() {
        this.status = "Completed";
        System.out.println("Trip has been completed.");
    }

    public int getId() {
        return id;
    }

    public String getPickupLocation() {
        return pickupLocation;
    }

    public String getDropOffLocation() {
        return dropOffLocation;
    }

    public RideType getRideType() {
        return rideType;
    }

    public double getFare() {
        return fare;
    }

    public double getDistance() {
        return distance;
    }

    public Driver getDriver() {
        return driver;
    }

    public String getStatus() {
        return status;
    }
}

File: Lab_1_RideSharingApp/Lab_1_RideSharingApp/Trip/Trip.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab_1_RideSharingApp
{
    public class Trip
    {
        public int Id { get; set; }
        public string PickupLocation { get; set; }
        public string DropOffLocation { get; set; }
        public RideType RideType { get; set; }
        public double Fare { get; private set; }
        public double Distance { get; set; }
        public Driver Driver { get; private set; }
        public string Status { get; private set; }

        public Trip(int id, string pickupLocation, string dropOffLocation, RideType rideType, double distance)
        {
            Id = id;
            PickupLocation = pickupLocation;
            DropOffLocation = dropOffLocation;
            RideType = rideType;
            Distance = distance;

            // Immediately calculate the fare after initialization
            Fare = CalculateFare();
            Status = "Requested";
        }

        public double CalculateFare()
        {
            // Updated fare calculation logic based on ride type
            double baseFare = 0;
            switch (RideType)
            {
                case RideType.Carpool:
                    baseFare = 20;
                    break;

                case RideType.Luxury:
                    baseFare = 35;
                    break;

                case RideType.Bike:
                    baseFare = 15;
                    break;

                default:
                    baseFare = 10; // Default fare for unknown ride types
                    break;
            }

            // Calculate fare based on the ride type and distance
            return baseFare * Distance;
        }

        public void AssignDriver(Driver driver)
        {
            Driver = driver;
            Status = "Driver Assigned";
        }

        public void StartTrip()
        {
            Status = "In Progress";
            Console.WriteLine("\nTrip has started.");
        }

        public void CompleteTrip()
        {
            Status = "Completed";
            Console.WriteLine("\nTrip has been completed.");
        }
    }

}
